### DISCUSSION TITLE
Add missing useful tools to PowerShell

### DISCUSSION THREAD

**Discussion by sba923** [PowerShell/PowerShell#15079] [General]
Hi everyone,

Coming from a long Korn Shell history (with a **lot** of (using, beta-testing, advocacy for) MKS in it @BrucePay 😋), I used to miss some key tools everyday, so I wrote crude versions (as scripts, not yet modules or anything fancy).

I'd like to use this conversation to:
- build a list of all such tools
- discuss on how they would be best added to PowerShell: built-in, module on the gallery...

Let's start with my current list, in no particular order:

1. `dirname`: enables oneliners such as `ii (dirname (dirname (gmo oh-my-posh).Path))`
2. `basename`
3. `whence`: for instance, makes it a snap to edit a script you've just run and had an issue:

`code (whence Set-WindowsTerminalProgressState)`

(where `code` resolves to my own `code.ps1`, but that's another discussion altogether)

4. `whatis`: outputs information about a command in a human-readable format, adding value on top of `Get-Command` in the process, such as version information for AppX reparse points:

````
PS❯ whatis whatis
'whatis' is an externalscript
Path is 's:\PowerShell\whatis.ps1'


PS❯ whatis swtps
'swtps' is an alias
Resolves to: 'Set-WindowsTerminalProgressState'


PS❯ whatis winget
'winget' is an application
Executable path is 'C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_1.11.10191.0_x64__8wekyb3d8bbwe\AppInstallerCLI.exe'
Description: 'AppInstallerCLI.exe'
Product version: '1.11.210119001-AppxClickHandlerstor'
Executable type: 64-bit


PS❯ whatis procexp
'procexp' is an application
Executable path is 'c:\usr\wbin\procexp.exe'
Description: 'Sysinternals Process Explorer'
Product version: '16.32'


PS❯ whatis runprocexp
'runprocexp' is a CMD script
c:\usr\wbin\runprocexp.cmd


PS❯ whatis sc
'sc' is an application
Executable path is 'C:\WINDOWS\system32\sc.exe'
Description: 'Service Control Manager Configuration Tool'
Product version: '10.0.19041.1'
Executable type: 64-bit


PS❯ whatis gc
'gc' is an alias
Resolves to: 'Get-Content'


PS❯ whatis Get-Content
'Get-Content' is a cmdlet
````

you get the idea

5. `df`: outputs human-readable information about local and network-mounted drives:

````

PS❯ df

DeviceID VolumeName         MountedOn        Size Freespace PercentFree PercentUsed
-------- ----------         ---------        ---- --------- ----------- -----------
C:       <REDACTED>                         857.9 GB  230.8 GB       26.90       73.10
D:       <REDACTED>_D                         1.8 TB  237.5 GB       12.75       87.25
E:
F:
G:
H:
I:       WD4TB_202103211456                3.6 TB    3.1 TB       85.11       14.89
L:       HGST4000-35                       3.6 TB    2.2 TB       61.19       38.81
S:       sto                \\pnjnas\sto   3.6 TB    1.4 TB       39.41       60.59
V:       V                  \\pnjnas\V     3.6 TB    1.4 TB       39.41       60.59
````

**Comment by SteveL-MSFT** [comment]
I think all of these so far could be in a module published to PowerShellGallery.  

`dirname` is basically `Split-Path $path -Parent`?
`basename` is basically `Split-Path $path -Leaf`?

For `whence`, it seems like maybe we should add a member to `FunctionInfo` with the source script?

`df` is certainly something that should exist, but can be a `Get-DiskFreespace` cmdlet that is cross-platform

One of the concerns we've had in the past is cmdlet aliases that overlap with existing Unix commands, so in this case, the module may only alias those functions if the native command doesn't exist, but then you have the problem where it's not switch compatible with the actual Unix commands.

**Reply by sba923** [reply]
Good remarks, esp. about the potential conflict on *nix with builtins (as is already the case for `ls`, `rm`...).



I realize I have said I have a crude (did I say _crude_?) implementation, but forgot to share it to fuel the discussion.

Here it is. Of course, improvement suggestions are welcome 😋.

[15079.zip](https://github.com/PowerShell/PowerShell/files/6178080/15079.zip)

For some reason I haven't "persisted" the implementation of `basename` I had written some time ago.

**Reply by sba923** [reply]
> For `whence`, it seems like maybe we should add a member to `FunctionInfo` with the source script?

I'm not sure to catch how this relates with my example / use case. Can you please clarify?

**Comment by sba923** [comment]
Any other thoughts? Is my next step to bite the bullet about writing modules and publish a first version? Any suggestion for the module's name?

**Reply by SeeminglyScience** [reply]
For stuff like this where it's mostly personalization and shortcuts, the profile is usually a good place for that. The way most folks publish their profiles is just creating a github repo called something like `dotfiles` and putting all their configs and profiles there.

**Reply by sba923** [reply]
That's a way... that's doesn't make the tools easy to discover.

@SteveL-MSFT what do _you_ think?

**Comment by sba923** [comment]
Created https://gist.github.com/sba923/d0d5ad16f2b12d7785adf830b0395dc2, FWIW

**Comment by kasini3000** [comment]
Here are some simple suggestions for powershell beginners:
3 code 'file full path'
4 get-command 'xxx'
5 Get-PhysicalDisk, Get-Volume # only win
Many other things can be implemented by writing a small script.

**Reply by sba923** [reply]
> Here are some simple suggestions for powershell beginners: 3 code 'file full path' 4 get-command 'xxx' 5 Get-PhysicalDisk, Get-Volume # only win Many other things can be implemented by writing a small script.

I'm not sure what you want to convey with that remark? That there are more helpers needed?

BTW I also have a `code.ps1` that allows me to invoke VScode from a PowerShell prompt in a number of scenarios: VScode not on PATH, argument is a relative or UNC path, etc.
